File: src/mtglab/cards/identify.py
"""Name the card in front of the camera, or offer what it might be.

The reader for the photographed deck (ROADMAP item 14). A capture hands this
module what it thought it saw -- a set code and a collector number off the
bottom-left corner, a title off the top, or both -- and gets back either a
card or a shortlist. What it never gets back is a guess.

**Two tiers, and the line between them is measured rather than assumed.**

A set code plus a collector number is a *lookup*. It picks one row out of
`printings` by what amounts to a compound key, there is no judgement in it,
and so a hit resolves outright.

A title is a *similarity*, and the numbers say a similarity must never
resolve anything. Simulating OCR damage over the 400 most-played cards, with
the character confusions a real engine makes (l/1, O/0, rn/m, cl/d):

    1 error : top-1 correct 120/120 | lowest CORRECT 0.875 | highest WRONG -
    2 errors: top-1 correct 116/120 | lowest CORRECT 0.801 | highest WRONG 0.917
    3 errors: top-1 correct 108/120 | lowest CORRECT 0.780 | highest WRONG 0.933
    5 errors: top-1 correct  72/120 | lowest CORRECT 0.740 | highest WRONG 0.942

The two distributions overlap, and they overlap *badly*: at three errors a
wrong card scores higher (0.933) than a right one does (0.780). No threshold
can separate them, so a threshold would only decide how often the app
asserts the wrong card confidently. It does not get to assert one at all --
`by_title` ranks, and a human picks. That is rule 4's ethic reached from a
measurement instead of a principle, and it is why `Reading.resolved` is
`None` for every title-only sighting no matter how good the score looks.

`CANDIDATES` is five for the same reason it is not eight: correct-answer-in-
the-list runs 99% / 95% / 80% at two, three and five errors, and is already
99% / 95% / 78% by the third. Past three the list stops paying.

**Front faces are matched too.** 889 names in the pool carry a `//`, and a
*flawlessly* read title of one loses without this: `Delver of Secrets` ranks
`Stealer of Secrets` (0.905) above `Delver of Secrets // Insectile
Aberration` (0.883). Scoring against the front face as well takes it to
1.000, because the camera can only ever see one side of the card.

**Banned cards are identifiable and non-cards are not.** The filter is
commander legality `IN ('legal', 'banned')`, which keeps Primeval Titan and
Emrakul -- two of this library's own decks run them -- while dropping the 87
emblems, 102 schemes, 207 planes and 107 vanguards that are in `oracle_cards`
and can never be in a deck. Refusing a banned card is the gate's job; the
reader's job is to say what is on the table.

Nothing here writes, and nothing here reaches a network.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any

#: How many names a title sighting offers. The knee is three; see the module
#: docstring for the curve this is read off.
CANDIDATES = 5

#: The most sightings one request may carry. A photo of a fanned spread holds
#: ten or so cards, and each title costs a scan of ~32k names (~16ms), so this
#: is a bound on work rather than a statement about how people use the camera.
MAX_SIGHTINGS = 40

#: Longest title text considered. The longest card name ever printed is 141
#: characters (the Unhinged elemental whose name is a joke about long card
#: names); past this it is not a misread title, it is a misread card.
MAX_TITLE = 200

#: Cards a Commander deck could contain, banned ones included. Interpolated
#: rather than parameterised because it carries no caller input.
_IS_CARD = "json_extract_string(legalities, 'commander') IN ('legal', 'banned')"

#: A collector number as the *card face* prints it: `284/281`, sometimes
#: zero-padded. The pool stores only the first half, so the denominator is
#: dropped before the lookup.
_FACE_NUMBER = re.compile(r"^\s*(?P<number>[^/\s]+)")

#: A set code is two to six alphanumerics, as `decklist.py` already has it.
#: Stored lowercase in the pool, which is why every comparison folds case.
_SET_CODE = re.compile(r"^[A-Za-z0-9]{2,6}$")


@dataclass(frozen=True)
class Sighting:
    """What one capture thought it saw.

    Every field is optional and independently unreliable, which is the whole
    reason this is three fields rather than a resolved card: the corner may
    be legible on a card whose title is glare, and a pre-2015 frame prints no
    collector number at all.
    """

    set_code: str | None = None
    collector_number: str | None = None
    title: str | None = None


@dataclass(frozen=True)
class Candidate:
    """A name the title *might* be, and how alike the two strings are.

    The score is a string distance and nothing more. It is carried so a
    client can order and shade a list, never so anything can threshold on it.
    """

    name: str
    score: float


@dataclass
class Reading:
    """What the pool made of one sighting."""

    #: A card name, and only ever from a printing lookup. `None` means the
    #: user still has to choose -- including when `candidates` looks certain.
    resolved: str | None = None
    #: How `resolved` was arrived at, or why nothing was: `printing`,
    #: `title` (offered, not decided), or `nothing`.
    via: str = "nothing"
    candidates: list[Candidate] = field(default_factory=list)


def face_number(text: str | None) -> str | None:
    """`284/281` -> `284`. The pool stores the numerator alone."""
    if not text:
        return None
    found = _FACE_NUMBER.match(text)
    return found["number"] if found else None


def by_printing(con: Any, set_code: str | None,
                collector_number: str | None) -> str | None:
    """The corner tier: one row out of `printings`, or nothing.

    Zero-padding is tolerated on both sides because the face pads and the
    pool does not (three of 12,650 stored numbers have a leading zero, and
    plenty of cards print `0284`).

    A `(set, number)` pair that somehow names two different cards resolves to
    neither -- that is an ambiguity, and this tier's whole claim is that it
    has no judgement in it.
    """
    number = face_number(collector_number)
    if not set_code or not number or not _SET_CODE.match(set_code):
        return None
    rows = con.execute(
        "SELECT DISTINCT name FROM printings "
        "WHERE lower(set_code) = lower(?) "
        "  AND (collector_number = ? "
        "       OR ltrim(collector_number, '0') = ltrim(?, '0')) "
        "LIMIT 2",
        [set_code, number, number],
    ).fetchall()
    return str(rows[0][0]) if len(rows) == 1 else None


def by_title(con: Any, title: str | None, *,
             limit: int = CANDIDATES) -> list[Candidate]:
    """The title tier: a ranked shortlist, and never an answer.

    Scored against the whole name *and* the front face, taking the better of
    the two, because the camera sees one side of a double-faced card.
    """
    if not title:
        return []
    text = title.strip()[:MAX_TITLE]
    if not text:
        return []
    rows = con.execute(
        "SELECT name, greatest("
        "         jaro_winkler_similarity(lower(name), lower(?)),"
        "         jaro_winkler_similarity("
        "           lower(split_part(name, ' // ', 1)), lower(?))) AS score "
        f"FROM oracle_cards WHERE {_IS_CARD} "
        "ORDER BY score DESC, edhrec_rank NULLS LAST "
        "LIMIT ?",
        [text, text, max(1, int(limit))],
    ).fetchall()
    return [Candidate(name=str(r[0]), score=float(r[1])) for r in rows]


def read(con: Any, sightings: list[Sighting]) -> list[Reading]:
    """Read a batch of captures, in the order they were taken.

    One `Reading` per `Sighting`, always, so a client can hold the two lists
    side by side; a sighting nothing could be made of comes back as
    `via="nothing"` rather than being dropped.
    """
    out: list[Reading] = []
    for sighting in sightings[:MAX_SIGHTINGS]:
        name = by_printing(con, sighting.set_code, sighting.collector_number)
        if name is not None:
            out.append(Reading(resolved=name, via="printing"))
            continue
        candidates = by_title(con, sighting.title)
        out.append(Reading(
            via="title" if candidates else "nothing",
            candidates=candidates,
        ))
    return out

File: go/internal/cards/identify.go
// Package cards is the reader for the photographed card (ADR 34's
// deterministic half) -- a set code and a collector number off the
// bottom-left corner is a *lookup* and resolves; a title is a *similarity*
// and only ever offers five names, because the measured scores of right and
// wrong answers overlap and no threshold can separate them. ADR 34 holds
// the argument.
package cards

import (
	"context"
	"regexp"
	"strings"

	"github.com/aasquier/sylvan-library/go/internal/pool"
)

// Candidates is how many names a title sighting offers; the knee is three.
const Candidates = 5

// MaxSightings bounds one request's work: a fanned spread is ten or so.
const MaxSightings = 40

// MaxTitle is the longest title considered; past it is a misread card.
const MaxTitle = 200

// isCard: cards a Commander deck could contain, banned ones included, and
// none of the emblems, schemes, planes and vanguards that cannot be in one.
const isCard = "json_extract_string(legalities, 'commander') IN ('legal', 'banned')"

var (
	faceNumber   = regexp.MustCompile(`^\s*([^/\s]+)`)
	setCodeRe    = regexp.MustCompile(`^[A-Za-z0-9]{2,6}$`)
	faceNumberIn = regexp.MustCompile(`(\d{1,4}[a-z]?)`)
	yearRe       = regexp.MustCompile(`^(19|20)\d\d$`)
	tokenSplit   = regexp.MustCompile(`[^0-9A-Za-z]+`)
)

// Sighting is what one capture thought it saw. Every field is optional and
// independently unreliable; "" is the absent value.
type Sighting struct {
	SetCode         string
	CollectorNumber string
	Title           string
	// The bottom-left block exactly as the reader saw it. Preferred over the
	// two fields above when present -- see FromCorner.
	Corner string
}

// Candidate is a name the title might be, and how alike the strings are.
// Carried so a client can order and shade a list, never so anything can
// threshold on it.
type Candidate struct {
	Name  string
	Score float64
}

// Reading is what the pool made of one sighting. `Resolved` is only ever a
// printing lookup's answer; `Via` is printing, title, or nothing.
type Reading struct {
	Resolved   string
	Via        string
	Candidates []Candidate
}

// SetCodes is every set code the pool knows, upper-cased. Read once per
// Read and handed down, never memoised on the package: the pool can be
// pointed elsewhere under a process.
func SetCodes(ctx context.Context, c *pool.Conn) (map[string]bool, error) {
	rows, err := c.DB().QueryContext(ctx,
		"SELECT DISTINCT upper(set_code) FROM printings WHERE set_code IS NOT NULL")
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := map[string]bool{}
	for rows.Next() {
		var code string
		if err := rows.Scan(&code); err != nil {
			return nil, err
		}
		out[code] = true
	}
	return out, rows.Err()
}

// FromCorner reads a set code and a collector number out of the raw corner
// block -- server-side because the answer needs the pool's 986 real set
// codes: the longest real prefix of a line's *first* token is the set code
// (the artist after it never gets a vote), and a token starting with a digit
// is the collector-number line, never a set code. A four-digit year is the
// copyright line bleeding into the crop.
func FromCorner(text string, codes map[string]bool) Sighting {
	if text == "" {
		return Sighting{Corner: text}
	}
	var number, code string
	lines := strings.Split(text, "\n")
	if len(lines) > 8 {
		lines = lines[:8]
	}
	for _, line := range lines {
		tokens := []string{}
		for _, t := range tokenSplit.Split(line, -1) {
			if t != "" {
				tokens = append(tokens, t)
			}
		}
		if len(tokens) == 0 {
			continue
		}
		if number == "" {
			for _, token := range tokens {
				if m := faceNumberIn.FindStringSubmatch(token); m != nil && !yearRe.MatchString(m[1]) {
					number = m[1]
					break
				}
			}
		}
		if code == "" {
			first := strings.ToUpper(tokens[0])
			if first[0] < '0' || first[0] > '9' {
				for size := 6; size > 1; size-- {
					if len(first) >= size && codes[first[:size]] {
						code = first[:size]
						break
					}
				}
			}
		}
	}
	return Sighting{SetCode: code, CollectorNumber: number, Corner: text}
}

// FaceNumber is `284/281` -> `284`: the pool stores the numerator alone.
func FaceNumber(text string) string {
	if text == "" {
		return ""
	}
	m := faceNumber.FindStringSubmatch(text)
	if m == nil {
		return ""
	}
	return m[1]
}

// ByPrinting is the corner tier: one row out of `printings`, or nothing.
// Zero-padding is tolerated on both sides; a pair naming two different cards
// resolves to neither, because this tier's whole claim is that it has no
// judgement in it.
func ByPrinting(ctx context.Context, c *pool.Conn, setCode, collectorNumber string) (string, error) {
	number := FaceNumber(collectorNumber)
	if setCode == "" || number == "" || !setCodeRe.MatchString(setCode) {
		return "", nil
	}
	rows, err := c.DB().QueryContext(ctx,
		`SELECT DISTINCT name FROM printings
		 WHERE lower(set_code) = lower(?)
		   AND (collector_number = ? OR ltrim(collector_number, '0') = ltrim(?, '0'))
		 LIMIT 2`, setCode, number, number)
	if err != nil {
		return "", err
	}
	defer rows.Close()
	names := []string{}
	for rows.Next() {
		var n string
		if err := rows.Scan(&n); err != nil {
			return "", err
		}
		names = append(names, n)
	}
	if err := rows.Err(); err != nil {
		return "", err
	}
	if len(names) == 1 {
		return names[0], nil
	}
	return "", nil
}

// ByTitle is the title tier: a ranked shortlist, and never an answer.
// Scored against the whole name *and* the front face, the better of the two,
// because the camera sees one side of a double-faced card.
func ByTitle(ctx context.Context, c *pool.Conn, title string, limit int) ([]Candidate, error) {
	out := []Candidate{}
	text := strings.TrimSpace(title)
	if text == "" {
		return out, nil
	}
	if runes := []rune(text); len(runes) > MaxTitle {
		text = string(runes[:MaxTitle])
	}
	if limit < 1 {
		limit = 1
	}
	rows, err := c.DB().QueryContext(ctx,
		`SELECT name, greatest(
		           jaro_winkler_similarity(lower(name), lower(?)),
		           jaro_winkler_similarity(lower(split_part(name, ' // ', 1)), lower(?))) AS score
		 FROM oracle_cards WHERE `+isCard+`
		 ORDER BY score DESC, edhrec_rank NULLS LAST
		 LIMIT ?`, text, text, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var cand Candidate
		if err := rows.Scan(&cand.Name, &cand.Score); err != nil {
			return nil, err
		}
		out = append(out, cand)
	}
	return out, rows.Err()
}

// Read reads a batch of captures, in the order they were taken: one Reading
// per Sighting, always, so a client can hold the two lists side by side.
func Read(ctx context.Context, c *pool.Conn, sightings []Sighting) ([]Reading, error) {
	batch := sightings
	if len(batch) > MaxSightings {
		batch = batch[:MaxSightings]
	}
	// One scan for the whole batch, and only when a corner needs reading.
	var codes map[string]bool
	for _, s := range batch {
		if s.Corner != "" && s.SetCode == "" {
			var err error
			if codes, err = SetCodes(ctx, c); err != nil {
				return nil, err
			}
			break
		}
	}
	out := make([]Reading, 0, len(batch))
	for _, raw := range batch {
		sighting := raw
		if raw.Corner != "" && raw.SetCode == "" {
			sighting = FromCorner(raw.Corner, codes)
		}
		name, err := ByPrinting(ctx, c, sighting.SetCode, sighting.CollectorNumber)
		if err != nil {
			return nil, err
		}
		if name != "" {
			out = append(out, Reading{Resolved: name, Via: "printing", Candidates: []Candidate{}})
			continue
		}
		candidates, err := ByTitle(ctx, c, raw.Title, Candidates)
		if err != nil {
			return nil, err
		}
		via := "nothing"
		if len(candidates) > 0 {
			via = "title"
		}
		out = append(out, Reading{Via: via, Candidates: candidates})
	}
	return out, nil
}
